pickhardtpayments/Attempt.py

```diff
@@ -0,0 +1,134 @@
+        for channel in path:
+            if not isinstance(channel, Channel):
+                raise ValueError("path needs to be a collection of Channels")
+            self._path = path
+        self._status = SettlementStatus.PLANNED
+        self._amount = amount
+
+    def __str__(self):
+        return "Path with {} channels to deliver {} sats and status {}.".format(len(self._path),
+                                                                                self._amount, self._status.name)
+
+    @property
+    def path(self):
+        """Returns the path of the attempt.
+
+        :return: the list of UncertaintyChannels that the path consists of
+        :rtype: list[UncertaintyChannel]
+        """
+        return self._path
+
+    @path.setter
+    def path(self, attempts: list):
+        """Sets the path that was set up for a certain amount
+
+        :param attempts: a List of Channels from UncertaintyGraph that an amount was attempted to be routed through
+        :type attempts: list[UncertaintyChannel]
+        """
+        if not isinstance(attempts, list):
+            raise ValueError("path needs to be a collection of Channels")
+        for attempt in attempts:
+            if not isinstance(attempt, Channel):
+                raise ValueError("path needs to be a collection of Channels")
+        self._paths = attempts
+
+    @property
+    def amount(self):
+        """Returns the amount of the attempt.
+
+        :return: the amount that was tried to send in this Attempt
+        :rtype: int
+        """
+        return self._amount
+
+    @property
+    def status(self):
+        """Returns the status of the attempt.
+
+        :return: returns the state of the attempt
+        :rtype: SettlementStatus
+        """
+        return self._status
+
+    @status.setter
+    def status(self, value: SettlementStatus):
+        """Sets the status of the attempt.
+
+        A flag to describe if the path failed, succeeded or was used for settlement (see enum SettlementStatus)
+
+        :param value: Current state of the Attempt
+        :type value: SettlementStatus
+        """
+        self._status = value
+
+    @property
+    def routing_fee(self):
+        """Returns the accrued routing fee in msat requested for this path.
+
+        :return: accrued routing fees for this attempt in msat
+        :rtype: int
+        """
+        return self._routing_fee
+
+    @routing_fee.setter
+    def routing_fee(self, value: int):
+        """Sets the accrued routing fee in msat requested for this path
+
+        :param value: accrued routing fees for this attempt in msat
+        :type value: int
+        """
+        self._routing_fee = value
+
+    @property
+    def probability(self):
+        """Returns estimated success probability before the attempt
+
+        :return: estimated success probability before the attempt
+        :rtype: float
+        """
+        return self._probability
```

Comment:
not entirely sure as the attempt class does not know the uncertainty network but we could compute this on the fly for the path (as the path contains such information)

Reply:
there is btw a strong attempt against it as non disjoint paths will change probability computation.

Reply:
yes, this is simply to collect statistical a priori information: 
In '_estimate_payment_statistics' this is added when calling `uncertainty_network.get_features_of_candidate_path. it's tied to this instance of attempt, so another Attempt on the same graph will have its own (then perhaps different probability). It should be purely descriptive and is not used for further calculation.

Reply:
### Bias in probabilities in UncertaintyGraph?

When collecting possible paths in a round/"payment loop", all path candidates are assumed to be sent and settled successfully. 
This is my assumption, because we iterate over each attempt: 
- 'UncertaintyChannel.success_probability' freshly calculates probabilities from current channel balances, and  
- then `allocate_amount_on_path` is updated before the loop commences with the next possibly not disjoint path, that then gets its probabilities from the just before adjusted inflight amount. 

_Or am I reading it wrong?_

Ist this something we can live with? 
What would be possible ways to correct for this (like trying to immediately figure out if the onion would fail or not)?

Reply:
This indeed the really **messy** book keeping that I introduced. I am happy to find a better solution to this! I need to allocate the amounts for planned onions to have probability correction of the next path correct. (Ask me for an example if you need one!) I remove all of the allocated HTLCs then before I do the actual sending / probing. It is really the most ugly part of my code :(